design/concise-tidy-design.Rmd

```diff
@@ -0,0 +1,200 @@
+Here is a first draft implementation of broom-like methods for `crmPack, as suggested in [issue 323](https://github.com/Roche/crmPack/issues/323).
+
+Whilst variants of "How do I convert an S4 object to a `data.frame`/`tibble`?" are commonly asked questions on StackOverflow [see, for example, [here](https://stackoverflow.com/questions/57317958/general-way-to-transform-s4-object-to-dataframe-r) and [here](https://stackoverflow.com/questions/14604269/how-to-create-a-dataframe-of-user-defined-s4-classes-in-r)], there appears to be no existing package to do this.  However, the approach generally suggested in answers to these questions is essentially the same as suggested here: 
+
+*  Deal with the object slot by slot  
+*  Convert each slot to a `tibble`  
+*  Return the slot-specific `tibble`s in a `list` 
+
+### Principles
+The general principles I propose for tidying `crmPack` objects are as follows:
+
+* All slots that are not functions are converted to `tibble`s or a list of `tibble`s.
+* If the slot is scalar, the slot is converted to a 1x1 `tibble`.  This will ease downstream operations such as `row-bind`ing.
+* If the object being tidied contains multiple slots of (potentially) different lengths, the results is a list of `tibble`s.  The list may be nested to multiple levels.  (See, for example, `LogisticLogNormal`.)
+* The column names of the tidied `tibble` correspond to the slot names of the parent object.
+* If an object contains a vector or matrix of values that correspond to an "obvious" set of parameters, columns take on the names of those parameters and the `tibble` is augmented by a column named `Parameter`.  This most often happens in the mean vector and covariance matrix of `ModelParamsNormal`.
+* When the value of a slot has not been set, a zero-row `tibble` is returned.
+* When the value of a slot has scalar attributes, these attributes are added as a column of the `tibble`, whose name is the name of the attribute and whose value is the value of the attribute for every row of the tibble.  Vector attributes will have to be handled on a case-by-case basis.
+* `tbl_<className>` is prepended to the class of the tidy `tibble`.
+
+### Exceptions
+*  Where a vector slot (or series of vector slots) define a range, then the naming convention described above is not followed.  Instead, columns named `min` and `max` define the extent of the range and a column named `Range` labels the range.
+
+### Please also read
+Please also read [Issue 407](https://github.com/Roche/crmPack/issues/407) regarding inconsistent slot names.
+
+## Implementation
+`tidy()` is a generic defined in `broom`.  An end-user of `crmPack` may not have installed `broom`.  Therefore, we have two options:
+
+* Make `crmPack` Depend on `broom`
+* On loading `crmPack`, determine if `broom` is an available namespace and, if not, define the appropriate generic
+
+After discussion, we take the first approach and make `crmPack` Depend on `broom`.
+
+### Simple examples
```

Comment:
also in terms of examples, it would be good to add the top level method, plus one specialization,
using helper functions across top level methods of different class hierarchies to avoid duplication of code

Reply:
Revised, hopefully as requested.